[nativert] AOTI lowering and packaging as NativeRT delegate

yiming0416 opened the pull request:
Summary:
A demo for creating AOTI delegate for NativeRT in OSS.

- It supports full graph lowering only.
- It leverages `executorch_call_delegate` HOP but doesn't rely on `executorch`.
- The delegate graph is obtained by tracing a `LoweredBackendModule` whose forward function calls `executorch_call_delegate`.
- The main difference between `executorch_call_delegate` and `aoti_call_delegate` is that the delegate graph from `executorch_call_delegate` doesn't have weights lifted as inputs.
- original_ep and delegate_ep are treated as flat EP dictionary and there is no nested structure.
- The naming contract is enforced by `model_name` and `backend_id`

Test Plan:
CI

Rollback Plan:

Differential Revision: D81641157

dolpm commented on torch/nativert/backends/lower_utils.py:
```diff
@@ -0,0 +1,50 @@
+            model_name: original_ep,
+            f"{model_name}-{backend_id}": delegate_ep,
+        },
+        aoti_files={f"{model_name}-{backend_id}": aoti_files},
```
nit: i feel this can be called delegate_files or something since we already know which backend they are relevant to from the id.

dolpm reviewed (approved):
address lints

angelayi commented on torch/_export/serde/serialize.py:
```diff
@@ -982,6 +982,15 @@ def serialize_input(self, arg, arg_type: Optional[Any] = None) -> Argument:
                     return Argument.create(
                         as_graph=GraphArgument(name=arg.target, graph=graph)
                     )
+                elif type(attr).__name__ == "LoweredBackendModule":
```
hmm I thought this would be a getattr node?

yiming0416 replied:
@angelayi  `executorch_call_delegate` hop only serves as an indicator for nativert to construct a AOTIDelegateExecutor in C++ runtime.

So all we need is a path in string format that can be looked up in the runtime for loading AOTI artifacts.

angelayi commented on torch/nativert/backends/lowered_aoti_module.py:
```diff
@@ -0,0 +1,30 @@
+class LoweredBackendModule(torch.nn.Module):
+    def __init__(
+        self,
+        module_name: str,
```
what is module_name?

yiming0416 replied:
@angelayi we use `{module_name}-{backend-id}` as the name of the delegate ep.
`module_name` is essentially the name of the model being lowered (given now we only support full graph lowering)

angelayi commented on torch/nativert/backends/lowered_aoti_module.py:
```diff
@@ -0,0 +1,30 @@
+        self,
+        module_name: str,
+        backend_id: str,
+        original_exported_program: ExportedProgram,
```
maybe we can try to make the signature consistent with [executorch's LoweredBackendModule](https://www.internalfb.com/code/fbsource/fbcode/executorch/exir/lowered_backend_module.py?lines=54)?

yiming0416 replied:
@angelayi Things like `CompileSpec` and `NamedDataStoreOutput` are exir-specific data structures.

The motivation of having this simplified `LoweredBackendModule` is that is doesn't depend on executorch and can be used standalone in `pytorch/pytorch`

yiming0416 committed:
[nativert] AOTI lowering and packaging as NativeRT delegate (#162285)

Summary:

A demo for creating AOTI delegate for NativeRT in OSS.

- It supports full graph lowering only.
- It leverages `executorch_call_delegate` HOP but doesn't rely on `executorch`.
- The delegate graph is obtained by tracing a `LoweredBackendModule` whose forward function calls `executorch_call_delegate`.
- The main difference between `executorch_call_delegate` and `aoti_call_delegate` is that the delegate graph from `executorch_call_delegate` doesn't have weights lifted as inputs.
- original_ep and delegate_ep are treated as flat EP dictionary and there is no nested structure.
- The naming contract is enforced by `model_name` and `backend_id`

Test Plan:
CI

Rollback Plan:

Reviewed By: dolpm

Differential Revision: D81641157
